### PR TITLE
drivers: interrupt_controller: Add SweRV PIC support

### PR DESCRIPTION
**Pull request by olofk**
Add support for the built-in Programmable Interrupt Controller
found in the SweRV EH1 RISC-V CPU

Signed-off-by: Olof Kindgren <olof.kindgren@gmail.com>

The driver is used in the [SweRVolf SoC](https://github.com/chipsalliance/Cores-SweRVolf)

### CODE REVIEW

**Comment by olofk** [comment]
Any takers on this? I understand it's a bit niche, but it has battle tested with a bunch of interrupt sources

**Review by kgugala — CHANGES_REQUESTED** [review]
@olofk, thanks for the contribution. I have some questions and remarks regarding the code.

File: drivers/interrupt_controller/swerv_pic.c [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+static void swerv_pic_init_core(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* No interrupts masked */
+	__asm__ swerv_pic_writecsr(meipt, 0);
+	__asm__ swerv_pic_writecsr(meicidpl, 0);
+	__asm__ swerv_pic_writecsr(meicurpl, 0);
+
+	swerv_pic_set_vector(swerv_pic_dummy);
```
[the remark this answers is not in the document]
Comment (olofk) [reply drivers/interrupt_controller/swerv_pic.c]: This was intended as initial support for the optional hardware vectored interrupts, but this feature is not yet used or tested, so it's better to remove it completely

File: drivers/interrupt_controller/swerv_pic.c [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+}
+
+void swerv_pic_dummy(void)
+{
+
+}
+
+static void swerv_pic_init_core(struct device *dev)
+{
+	ARG_UNUSED(dev);
```
[the remark this answers is not in the document]
Comment (olofk) [reply drivers/interrupt_controller/swerv_pic.c]: With the set_vector function removed I folded the logic into `swerv_pic_init` and removed this function

File: drivers/interrupt_controller/swerv_pic.c [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+	}
+
+	/* clear pending of each interrupt line */
+	for (i = 1; i < SWERV_PIC_MAX_ID; i++) {
+		swerv_pic_write(SWERV_PIC_meigwclr(i), 0);
+	}
+
+	swerv_pic_init_core(dev);
+
+	/* Setup IRQ handler for PLIC driver */
```
[the remark this answers is not in the document]
Comment (olofk) [reply drivers/interrupt_controller/swerv_pic.c]: Yes. Fixed now

File: drivers/interrupt_controller/swerv_pic.c [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+	IRQ_CONNECT(RISCV_MACHINE_EXT_IRQ,
+		    0,
+		    swerv_pic_irq_handler,
+		    NULL,
+		    0);
+
+	/* Enable IRQ for PLIC driver */
```
[the remark this answers is not in the document]
Comment (olofk) [reply drivers/interrupt_controller/swerv_pic.c]: Yes. Fixed now

**Comment by olofk** [comment]
Many thanks for the review @kgugala . I have fixed the issues you reported, rebased against master and pushed a new version. The following changes are also made to the new version

* Drop the logging code. It seems no other interrupt drivers have logging and I guess it's a bit awkward to log inside an interrupt routine anyway
* Converted all `z_arch` to `arch` to follow the API namespace changes

File: drivers/interrupt_controller/swerv_pic.c [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+	return swerv_pic_read(SWERV_PIC_meie(irq - RISCV_MAX_GENERIC_IRQ))
+	  & 0x1;
+}
+
+void swerv_pic_set_priority(u32_t irq, u32_t priority)
+{
+	u32_t key;
+
+	if (irq <= RISCV_MAX_GENERIC_IRQ)
```
Comment (kgugala) [inline drivers/interrupt_controller/swerv_pic.c]: please add braces here https://docs.zephyrproject.org/latest/contribute/index.html#coding-style

**Review by kgugala — CHANGES_REQUESTED** [review]
Small formatting comment, otherwise LGTM

File: drivers/interrupt_controller/swerv_pic.c (hunk 5 above)
Comment (olofk) [reply drivers/interrupt_controller/swerv_pic.c]: Thanks. Looks like both checkpatch and I missed that one. Fixed, rebased and pushed

**Review by kgugala — APPROVED** [review]

**Comment by olofk** [comment]
@npitre @ioannisg @andrewboie @galak @nategraff-sifive  Any updates on this?

**Comment by npitre** [comment]
On Wed, 4 Dec 2019, Olof Kindgren wrote:

> @npitre @ioannisg @andrewboie @galak @nategraff-sifive  Any updates on this?

I do not have experience with Zephyr IRQ driver so I don't have much to contribute here.

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**